plugins/hls-explicit-record-fields-plugin/src/Ide/Plugin/ExplicitFields.hs

```diff
@@ -0,0 +1,266 @@
+renderRecordInfo (RecordInfoPat ss pat) = RenderedRecordInfo ss <$> showRecordPat pat
+renderRecordInfo (RecordInfoCon ss expr) = RenderedRecordInfo ss <$> showRecordCon expr
+
+-- `Outputable` instance of `HsRecFields` does smart things to print
+-- the records that originally had wildcards with dots, even after they
+-- are removed by the renamer pass. Here `rec_dotdot` is set to
+-- `Nothing` so that fields are printed without such post-processing.
+preprocessRecord :: HsRecFields (GhcPass c) arg -> HsRecFields (GhcPass c) arg
+preprocessRecord flds = flds { rec_dotdot = Nothing , rec_flds = rec_flds' }
+  where
+    no_pun_count = maybe (length (rec_flds flds)) unLoc (rec_dotdot flds)
+    -- Field binds of the explicit form (e.g. `{ a = a' }`) should be
+    -- left as is, hence the split.
+    (no_puns, puns) = splitAt no_pun_count (rec_flds flds)
+    -- `hsRecPun` is set to `True` in order to pretty-print the fields as field
+    -- puns (since there is similar mechanism in the `Outputable` instance as
+    -- explained above).
+    puns' = map (mapLoc (\fld -> fld { hfbPun = True })) puns
+    rec_flds' = no_puns <> puns'
+
+showRecordPat :: Outputable (Pat (GhcPass c)) => Pat (GhcPass c) -> Maybe Text
+showRecordPat pat@(ConPat _ _ (RecCon flds)) =
+  Just $ printOutputable $
+    pat { pat_args = RecCon (preprocessRecord flds) }
+showRecordPat _ = Nothing
+
+showRecordCon :: Outputable (HsExpr (GhcPass c)) => HsExpr (GhcPass c) -> Maybe Text
+showRecordCon expr@(RecordCon _ _ flds) =
+  Just $ printOutputable $
+    expr { rcon_flds = preprocessRecord flds }
+showRecordCon _ = Nothing
+
+collectRecords :: GenericQ [RecordInfo]
+collectRecords = everything (<>) (maybeToList . (Nothing `mkQ` getRecPatterns `extQ` getRecCons))
+
+getRecCons :: LHsExpr (GhcPass 'Renamed) -> Maybe RecordInfo
+getRecCons e@(unLoc -> RecordCon _ _ flds)
+  | isJust (rec_dotdot flds) = Just $ mkRecInfo e
+  where
+    mkRecInfo :: LHsExpr (GhcPass 'Renamed) -> RecordInfo
+    mkRecInfo expr = RecordInfoCon (getLoc expr) (unLoc expr)
+getRecCons _ = Nothing
+
+getRecPatterns :: LPat (GhcPass 'Renamed) -> Maybe RecordInfo
+getRecPatterns conPat@(unLoc -> ConPat _ _ (RecCon flds))
+  | isJust (rec_dotdot flds) = Just $ mkRecInfo conPat
+  where
+    mkRecInfo :: LPat (GhcPass 'Renamed) -> RecordInfo
+    mkRecInfo pat = RecordInfoPat (getLoc pat) (unLoc pat)
+getRecPatterns _ = Nothing
+
+collectRecords' :: MonadIO m => IdeState -> NormalizedFilePath -> ExceptT String m CollectRecordsResult
+collectRecords' ideState =
+  handleMaybeM "Unable to TypeCheck"
+    . liftIO
+    . runAction "ExplicitFields" ideState
+    . use CollectRecords
+
+collectRecordsInRange :: MonadIO m => Range -> IdeState -> NormalizedFilePath -> ExceptT String m CollectRecordsResult
+collectRecordsInRange range ideState nfp = do
+  CRR renderedRecs exts <- collectRecords' ideState nfp
+  pure $ CRR (filter inRange renderedRecs) exts
```

Comment:
Just thinking through how this works: we collect all the record stuff in the module and cache it with a rule, good. But then we do a linear amount of work every time we get a code action request filtering them to see if they're in range. Maybe that's fine, but I could imagine this blowing up in modules with a lot of record stuff.

A cheap improvement would be to do some kind of binary search for the range filtering. Have a look at the code-ranges plugin, which I think does something similar.

Reply:
I have made some changes to the filtering logic: Now the rule builds and caches an `IntervalMap` rather than a list, so the filtering should be more efficient. Can you take a look to see if the changes are reasonable?

Reply:
Looks like `hw-prim` package (transitively depended on by `hw-fingertree`) doesn't yet work with GHC 9.4 though.